### PR TITLE
Feat eldad4 coroutines

### PR DESCRIPTION
**Pull request by eldadfux**
## What does this PR do?

(Provide a description of what this PR does and why it's needed.)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Screenshots may also be helpful.)

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [ ] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/controllers/api/avatars.php [hunk 1]
```diff
@@ -594,8 +595,9 @@
     ->inject('contributors')
     ->inject('employees')
     ->inject('logger')
-    ->action(function (string $userId, string $mock, int $width, int $height, Document $user, Document $project, Database $dbForProject, Database $dbForConsole, Response $response, array $heroes, array $contributors, array $employees, ?Logger $logger) use ($getUserGitHub) {
-        $user = Authorization::skip(fn () => $dbForConsole->getDocument('users', $userId));
+    ->inject('authp')
```
Comment (christyjacob4) [inline app/controllers/api/avatars.php:598]:
```suggestion
    ->inject('auth')
```

File: app/init2.php [hunk 2]
```diff
@@ -0,0 +1,1371 @@
...
+            $jwtUserId = $payload['userId'] ?? '';
+            $jwtSessionId = $payload['sessionId'] ?? '';
+
+            if ($jwtUserId && $jwtSessionId) {
+                $user = $dbForProject->getDocument('users', $jwtUserId);
+            }
+
+            if (empty($user->find('$id', $jwtSessionId, 'sessions'))) { // Match JWT to active token
+                $user = new Document([]);
+            }
```
Comment (stnguyen90) [inline app/init2.php:644]: There's a long-standing [problem](https://github.com/appwrite/appwrite/issues/8174) where calling `account.getSession('current')` doesn't work with a JWT because we look up a session using the session secret even though the JWT has the session id. 

Can we do what [this](https://github.com/appwrite/appwrite/pull/8313) PR does to fix the problem by adding sessionId to `Authentication` and keeping track of it?

File: app/init2.php [hunk 3]
```diff
@@ -0,0 +1,1371 @@
...
+
+            $jwtUserId = $payload['userId'] ?? '';
+            $jwtSessionId = $payload['sessionId'] ?? '';
+
+            if ($jwtUserId && $jwtSessionId) {
+                $user = $dbForProject->getDocument('users', $jwtUserId);
+            }
+
+            if (empty($user->find('$id', $jwtSessionId, 'sessions'))) { // Match JWT to active token
```
Comment (stnguyen90) [inline app/init2.php:642]: There's [another bug](https://github.com/appwrite/appwrite/issues/8000) where the JWT is still considered valid even though the session is expired. Can we fix that too like [this](https://github.com/appwrite/appwrite/pull/8313) PR?

**Comment by aabmets** [comment]
@stnguyen90 May I please ask a clarification about which endpoint performances are being measured here? Is this the performance of function workers, or requests to the Appwrite API in general?

> ## ✨ Benchmark results
> * Requests per second: 3,646
> * Requests with 200 status code: 656,397
> * P99 latency: 0.069279042
> 
> ## ⚡ Benchmark Comparison
> Metric	This PR	Latest version
> RPS	3,646	2,219
> 200	656,397	399,561
> P99	0.069279042	0.0737149

File: src/Appwrite/Specification/Format/Swagger2.php [hunk 4]
```diff
@@ -270,8 +270,14 @@ public function parse(): array
             );
 
             foreach ($parameters as $name => $param) { // Set params
+                $injections = [];
+
+                if (isset($param['injections'])) {
+                    $injections = array_map(fn ($injection) => $this->http->getContainer()->get($injection), $param['injections']);
+                }
```
Comment (christyjacob4) [inline src/Appwrite/Specification/Format/Swagger2.php:277]:
```suggestion
          $injections = array_map(fn ($injection) => $this->http->getContainer()->get($injection), $param['injections'] ?? []);
```

Maybe? 
I would also generate the spec file after this change and compare the current latest to make sure there is no strange diff 💡

File: src/Appwrite/Specification/Format/OpenAPI3.php [hunk 5]
```diff
@@ -269,10 +272,14 @@ public function parse(): array
...
             foreach ($route->getParams() as $name => $param) { // Set params
-                /**
-                 * @var \Utopia\Validator $validator
-                 */
-                $validator = (\is_callable($param['validator'])) ? call_user_func_array($param['validator'], $this->app->getResources($param['injections'])) : $param['validator'];
+                $injections = [];
+
+                if (isset($param['injections'])) {
+                    $injections = array_map(fn ($injection) => $this->http->getContainer()->get($injection), $param['injections']);
+                }
```
Comment (christyjacob4) [inline src/Appwrite/Specification/Format/OpenAPI3.php:279]:
```suggestion
                $injections = array_map(fn ($injection) => $this->http->getContainer()->get($injection), $param['injections'] ?? []);
```

File: app/cli.php [hunk 6]
```diff
@@ -7,209 +7,130 @@
...
-                ->setMetadata('project', 'console');
-
-            // Ensure tables exist
-            $collections = Config::getParam('collections', [])['console'];
-            $last = \array_key_last($collections);
-
-            if (!($dbForConsole->exists($dbForConsole->getDatabase(), $last))) { /** TODO cache ready variable using registry */
-                throw new Exception('Tables not ready yet.');
-            }
+global $global, $container;
```
Comment (christyjacob4) [inline app/cli.php:21]:
```suggestion
global $registry, $container;
```
can we just use registry?

File: app/cli.php [hunk 7]
```diff
@@ -7,209 +7,130 @@
...
-        Console::warning("Failed: {$error->getMessage()}");
-        Console::warning($error->getTraceAsString());
-    };
-}, ['register']);
+            Console::warning("Failed: {$error->getMessage()}");
+            Console::warning($error->getTraceAsString());
+        };
+    });
+
+$container->set($context);
```
Comment (christyjacob4) [inline app/cli.php:108]: what about other resources we removed like dbforproject? and some queues?

File: app/controllers/api/avatars.php [hunk 8]
```diff
@@ -154,11 +156,42 @@
             'id' => $githubId
         ];
     } catch (Exception $error) {
+        if ($logger) {
+            $version = System::getEnv('_APP_VERSION', 'UNKNOWN');
+
+            $log = new Log();
```
Comment (christyjacob4) [inline app/controllers/api/avatars.php:162]: why's this code here? We have global handlers for error right?

File: app/cli.php (hunk 7 above)
Comment (byawitz) [reply app/cli.php:108]: They are set in the `init.php` file

File: tests/unit/GraphQL/BuilderTest.php [hunk 9]
```diff
@@ -6,23 +6,25 @@
...
+        Mapper::init(Response\Models::getModels());
     }
 
     /**
      * @throws \Exception
      */
     public function testCreateTypeMapping()
     {
-        $model = $this->response->getModel(Response::MODEL_COLLECTION);
+        $model = Response\Models::getModel(Response::MODEL_COLLECTION);
```
Comment (christyjacob4) [inline tests/unit/GraphQL/BuilderTest.php:27]: can we just do `Models::` ?

File: tests/unit/GraphQL/BuilderTest.php [hunk 10]
```diff
@@ -6,23 +6,25 @@
...
 class BuilderTest extends TestCase
 {
     protected ?Response $response = null;
 
     public function setUp(): void
     {
-        $this->response = new Response(new SwooleResponse());
-        Mapper::init($this->response->getModels());
+        Response\Models::init();
+        $this->response = new Response(new UtopiaSwooleResponse(new SwooleResponse()));
```
Comment (christyjacob4) [inline tests/unit/GraphQL/BuilderTest.php:18]: this looks messy 😬  response(response(response(😵‍💫)))

File: app/controllers/api/avatars.php (hunk 8 above)
Comment (byawitz) [reply app/controllers/api/avatars.php:162]: This is an old code implantation. Maybe because it's being called inside a controller and we don't want the effect to propagate to the end user

File: tests/unit/Event/EventTest.php [hunk 11]
```diff
@@ -66,9 +66,13 @@ public function testParams(): void
         $this->assertEquals('eventValue1', $this->object->getParam('eventKey1'));
         $this->assertEquals('eventValue2', $this->object->getParam('eventKey2'));
         $this->assertEquals(null, $this->object->getParam('eventKey3'));
-        global $register;
-        $pools = $register->get('pools');
-        $client = new Client($this->object->getQueue(), $pools->get('queue')->pop()->getResource());
+
+        global $global;
```
Comment (christyjacob4) [inline tests/unit/Event/EventTest.php:70]: why did we change the name from register / registry to $global? its confusing especially since we have a keyword with similar name

File: tests/resources/docker/docker-compose.yml [hunk 12]
```diff
@@ -324,7 +324,7 @@ services:
       - MYSQL_USER=user
       - MYSQL_PASSWORD=password
       - MARIADB_AUTO_UPGRADE=1
-    command: 'mysqld --innodb-flush-method=fsync'
+    command: 'mysqld --innodb-flush-method=fsync --max_connections=5000'
```
Comment (christyjacob4) [inline tests/resources/docker/docker-compose.yml:327]: do we need to make similar changes to allow more connections on cloud?

File: tests/e2e/Services/Databases/DatabasesBase.php [hunk 13]
```diff
@@ -2868,7 +2868,14 @@ public function testInvalidDocumentStructure()
...
         $this->assertEquals('Invalid document structure: Attribute "probability" has invalid format. Value must be a valid range between 0 and 1', $badProbability['body']['message']);
         $this->assertEquals('Invalid document structure: Attribute "upperBound" has invalid format. Value must be a valid range between -9,223,372,036,854,775,808 and 10', $tooHigh['body']['message']);
-        $this->assertEquals('Invalid document structure: Attribute "lowerBound" has invalid format. Value must be a valid range between 5 and 9,223,372,036,854,775,807', $tooLow['body']['message']);
+
+        $max = '9,223,372,036,854,775,807';
+
+        if (PHP_VERSION_ID < 80300) {
+            $max = '9,223,372,036,854,775,808';
+        }
+
```
Comment (christyjacob4) [inline tests/e2e/Services/Databases/DatabasesBase.php:2877]: why this check? cant we just use the PHP_INT_MAX constant or so?

Comment (byawitz) [reply tests/e2e/Services/Databases/DatabasesBase.php:2877]: It's a PHP < 8.3 bug to have different value for the final number_format function
image

File: src/Appwrite/Utopia/Response.php [hunk 14]
```diff
@@ -605,7 +306,7 @@ public function dynamic(Document $document, string $model): void
     public function output(Document $document, string $model): array
     {
         $data       = clone $document;
-        $model      = $this->getModel($model);
+        $model      = Response\Models::getModel($model);
```
Comment (christyjacob4) [inline src/Appwrite/Utopia/Response.php:309]:
```suggestion
        $model      = Models::getModel($model);
```
just Models?

File: src/Appwrite/GraphQL/Schema.php [hunk 15]
```diff
@@ -234,23 +245,23 @@ protected static function collections(
...
                         $params['update'],
-                    )
+                    )*/
                 ];
                 $mutationFields[$collectionId . 'Delete'] = [
                     'type' => Mapper::model('none'),
                     'args' => Mapper::args('id'),
-                    'resolve' => Resolvers::documentDelete(
-                        $utopia,
+                    /*'resolve' => Resolvers::documentDelete(
```
Comment (christyjacob4) [inline src/Appwrite/GraphQL/Schema.php:259]: no resolve function?

File: tests/e2e/Services/Databases/DatabasesBase.php (hunk 13 above)
Comment (byawitz) [reply tests/e2e/Services/Databases/DatabasesBase.php:2877]: Guess it can be changed to this
```php
$max = \number_format(PHP_INT_MAX);
```

File: src/Appwrite/Auth/Authentication.php [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+class Authentication
+{
+    /**
+     * User Unique ID.
+     *
+     * @var string
+     */
+    private $unique = '';
+
```
Comment (christyjacob4) [inline src/Appwrite/Auth/Authentication.php:13]: what is the function of this class? Why did we remove it from Auth? The name Authentication doesnt reflect what this class does?

File: mariadb-config.cnf [hunk 17]
```diff
@@ -0,0 +1,2 @@
+[mysqld]
+max_connections=1024
```
Comment (christyjacob4) [inline mariadb-config.cnf:2]: do we need this? it can be done through docker compose cli args itself right?

File: phpstan.neon [hunk 18]
```diff
@@ -0,0 +1,11 @@
...
+  level: 0
+  scanDirectories:
+    - vendor/swoole/ide-helper
+  excludePaths:
+    - tests/resources
+  ignoreErrors:
+    - '#Parameter \$geodb of anonymous function has invalid type MaxMind\\Db\\Reader\.#'
+    - '#Parameter \$geodb of function router\(\) has invalid type MaxMind\\Db\\Reader\.#'
+    - '#Instantiated class MaxMind\\Db\\Reader not found.#'
+    - '#Function scrypt not found\.#'
```
Comment (christyjacob4) [inline phpstan.neon:11]: Are we using phpstan? or this is a leftover?

File: src/Appwrite/GraphQL/Schema.php (hunk 15 above)
Comment (byawitz) [reply src/Appwrite/GraphQL/Schema.php:259]: The [collection ](https://github.com/appwrite/appwrite/blob/5d70884e6cf4aa65e4da38f1bff870d4725964f0/src/Appwrite/GraphQL/Schema.php#L54) part was commented before, that's the only way to access that functions so they not in use right now, as of that they got commented out

File: src/Appwrite/Auth/Auth.php [hunk 19]
```diff
@@ -91,37 +91,6 @@ class Auth
...
-     * Set Cookie Name.
-     *
-     * @param $string
-     *
-     * @return string
-     */
-    public static function setCookieName($string)
-    {
-        return self::$cookieName = $string;
-    }
```
Comment (christyjacob4) [inline src/Appwrite/Auth/Auth.php:123]: not sure if removing it from here is a good idea? what was the reason?

File: src/Appwrite/Auth/Authentication.php (hunk 16 above)
Comment (byawitz) [reply src/Appwrite/Auth/Authentication.php:13]: This class holds the user's Cookie and secret.
Before that the details were part of the Auth class and they were static. meaning if two users are accessing the server in the same worker using different coroutines, they will share those details

File: phpstan.neon (hunk 18 above)
Comment (byawitz) [reply phpstan.neon:11]: We are

File: src/Appwrite/Auth/Auth.php (hunk 19 above)
Comment (byawitz) [reply src/Appwrite/Auth/Auth.php:123]: It's static and will be shared between coroutines check [this](https://github.com/appwrite/appwrite/pull/7975#discussion_r1745942848)

File: app/controllers/general.php [hunk 20]
```diff
@@ -713,7 +733,7 @@ function router(App $utopia, Database $dbForConsole, callable $getProjectDB, Swo
         }
 
         switch ($class) {
-            case 'Utopia\Exception':
+            case 'Utopia\Servers\Exception':
```
Comment (christyjacob4) [inline app/controllers/general.php:736]: hmm was the namespace changed?

Comment (byawitz) [reply app/controllers/general.php:736]: Yes, but it should be `Utopia\Http\Exception`

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php [hunk 21]
```diff
@@ -68,8 +68,11 @@ protected function enqueueResources(Group $pools, Database $dbForConsole): void
             \go(function () use ($delay, $scheduleKeys, $pools) {
                 \sleep($delay); // in seconds
 
-                $queue = $pools->get('queue')->pop();
-                $connection = $queue->getResource();
+                $pool = $pools['pools-queue-queue']['pool'];
+                $connection = $pool->get();
+                $this->connections->add($connection, $pool);
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/ScheduleFunctions.php:73]: and what is this->connections? Whats it used for?

Comment (byawitz) [reply src/Appwrite/Platform/Tasks/ScheduleFunctions.php:73]: check [this](https://github.com/appwrite/appwrite/pull/7975#discussion_r1745904413)

File: app/controllers/shared/api.php [hunk 22]
```diff
@@ -794,10 +810,16 @@
...
         if (System::getEnv('_APP_USAGE_STATS', 'enabled') !== 'enabled') {
             throw new Exception(Exception::GENERAL_USAGE_DISABLED);
         }
     });
+
+Http::shutdown()
+    ->inject('connections')
+    ->action(function (Connections $connections) {
+        $connections->reclaim();
+    });
```
Comment (christyjacob4) [inline app/controllers/shared/api.php:825]: where was this code before? I guess in http.php right? any reasons  for moving it here?

File: app/controllers/general.php [hunk 23]
```diff
@@ -1028,16 +1023,36 @@ function router(App $utopia, Database $dbForConsole, callable $getProjectDB, Swo
...
     });
 
 foreach (Config::getParam('services', []) as $service) {
-    include_once $service['controller'];
+    //include_once $service['controller'];
 }
+
+include_once 'mock.php';
+include_once 'shared/api.php';
+include_once 'shared/api/auth.php';
```
Comment (christyjacob4) [inline app/controllers/general.php:1039]: why do we need to do this now? Previously we didnt need to

File: app/controllers/general.php [hunk 24]
```diff
@@ -922,63 +909,71 @@ function router(App $utopia, Database $dbForConsole, callable $getProjectDB, Swo
...
         $mainDomain = System::getEnv('_APP_DOMAIN', '');
 
         if ($host === $mainDomain || $host === 'localhost') {
             $template = new View(__DIR__ . '/../views/general/humans.phtml');
             $response->text($template->render(false));
         } else {
-            router($utopia, $dbForConsole, $getProjectDB, $swooleRequest, $request, $response, $queueForEvents, $queueForUsage, $geodb);
+            if (is_null($route)) {
+                $route = new Route($request->getMethod(), $request->getURI());
+            }
```
Comment (christyjacob4) [inline app/controllers/general.php:971]: why would it be null ? it should be set in one of the ::init() hooks right? or automatically inside framework?

File: app/controllers/shared/api.php (hunk 22 above)
Comment (byawitz) [reply app/controllers/shared/api.php:825]: It's a new connections check [this](https://github.com/appwrite/appwrite/pull/7975#discussion_r1745904413)

File: src/Appwrite/Auth/Validator/MockNumber.php [hunk 25]
```diff
@@ -46,7 +46,7 @@ public function isValid($value): bool
             return false;
         }
 
-        $otp = new Text(6, 6, Text::NUMBERS);
+        $otp = new Text(6, 6, Validator\Text::NUMBERS);
```
Comment (christyjacob4) [inline src/Appwrite/Auth/Validator/MockNumber.php:49]:
```suggestion
        $otp = new Text(6, 6, Text::NUMBERS);
```
Can be removed right?

File: app/worker.php [hunk 26]
```diff
@@ -2,280 +2,109 @@
...
+$container->set($auditRetention);
+$container->set($deviceForCache);
+$container->set($queueForUsageDump);
+$container->set($executionRetention);
+$container->set($deviceForLocalFiles);
 
-$pools = $register->get('pools');
 $platform = new Appwrite();
-$args = $platform->getEnv('argv');
+$args = $_SERVER['argv'];
```
Comment (christyjacob4) [inline app/worker.php:107]: We created these wrapper methods in the platform library like getEnv inorder to avoid interacting with the SERVER array directly. We should use the getEnv() method here

File: app/realtime.php [hunk 27]
```diff
@@ -638,7 +586,7 @@ function getCache(): Cache
             $server->close($connection, $th->getCode());
         }
     } finally {
-        $register->get('pools')->reclaim();
+        //       TODO NOW $registry->get('pools')->reclaim();
```
Comment (christyjacob4) [inline app/realtime.php:589]: how does the reclaim happen here?

File: app/controllers/shared/api/auth.php [hunk 28]
```diff
@@ -46,15 +48,17 @@
...
-        $isPrivilegedUser = Auth::isPrivilegedUser(Authorization::getRoles());
-        $isAppUser = Auth::isAppUser(Authorization::getRoles());
+        $isPrivilegedUser = Auth::isPrivilegedUser($authorization->getRoles());
+        $isAppUser = Auth::isAppUser($authorization->getRoles());
 
         if ($isAppUser || $isPrivilegedUser) { // Skip limits for app and console devs
             return;
         }
 
+        if (str_contains($route->getPath(), '/v1/graphql')) { // Skip for graphQL recursive call
```
Comment (christyjacob4) [inline app/controllers/shared/api/auth.php:58]: this check doesnt feel safe. Maybe use `route->getLabel('sdk.namespace') === 'graphql')`

File: app/controllers/general.php [hunk 29]
```diff
@@ -1028,16 +1019,36 @@ function router(App $utopia, Database $dbForConsole, callable $getProjectDB, Swo
...
-
-App::wildcard()
+Http::wildcard()
     ->groups(['api'])
     ->label('scope', 'global')
     ->action(function () {
         throw new AppwriteException(AppwriteException::GENERAL_ROUTE_NOT_FOUND);
     });
 
 foreach (Config::getParam('services', []) as $service) {
```
Comment (christyjacob4) [inline app/controllers/general.php:1029]: I think this foreach can be removed now that we manually include all the files below?

File: app/controllers/general.php [hunk 30]
```diff
@@ -759,34 +778,8 @@ function router(App $utopia, Database $dbForConsole, callable $getProjectDB, Swo
             $publish = $error->getCode() === 0 || $error->getCode() >= 500;
         }
 
-        if ($error->getCode() >= 400 && $error->getCode() < 500) {
```
Comment (christyjacob4) [inline app/controllers/general.php:762]: I think we should keep this as we still use the expermental logger

File: app/init.php [hunk 31]
```diff
@@ -791,205 +233,172 @@ function (mixed $value) {
...
+                'multiple' => true,
+                'schemes' => ['redis'],
+            ],
+        ];
 
-            if (!in_array($dsnScheme, $schemes)) {
-                throw new Exception(Exception::GENERAL_SERVER_ERROR, "Invalid console database scheme");
-            }
+        $pools = [];
+        $poolSize = (int)System::getEnv('_APP_POOL_SIZE', 64);
```
Comment (christyjacob4) [inline app/init.php:310]: lets make sure we also document this in the variables .php file and deprecate the old variables. 
And update the docker compose and docker compose.phtml files to use the new POOL_SIZE variable

File: app/init.php [hunk 32]
```diff
@@ -791,205 +233,172 @@ function (mixed $value) {
...
+                                ->withCharset('utf8mb4')
+                                ->withUsername($dsnUser)
+                                ->withPassword($dsnPass)
+                                ->withOptions([
+                                    // No need to set PDO::ATTR_ERRMODE it is overwitten in PDOProxy
+                                    // PDO::ATTR_TIMEOUT => 3, // Seconds
+                                    // PDO::ATTR_PERSISTENT => true,
+                                    PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
+                                    PDO::ATTR_EMULATE_PREPARES => true,
+                                    PDO::ATTR_STRINGIFY_FETCHES => true,
```
Comment (christyjacob4) [inline app/init.php:365]: previously these were present
PDO::ATTR_TIMEOUT => 3, // Seconds
PDO::ATTR_PERSISTENT => true,

and PDO::MYSQL_ATTR_USE_BUFFERED_QUERY => true, was not there.. 
are these changes required?

File: src/Appwrite/Auth/Authentication.php (hunk 16 above)
Comment (christyjacob4) [reply src/Appwrite/Auth/Authentication.php:13]: We probably need a better name to reflect its purpose

**Comment by byawitz** [comment]
@aabmets, you can find the benchmark details [here](https://github.com/appwrite/appwrite/blob/1.6.x/.github/workflows/tests.yml#L152)

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php [hunk 33]
```diff
@@ -22,11 +21,16 @@ public static function getSupportedResource(): string
...
     {
-        $queue = $pools->get('queue')->pop();
-        $connection = $queue->getResource();
-        $queueForFunctions = new Func($connection);
+        [$connection,$pool, $dbForConsole] = $getConsoleDB();
+        $this->connections->add($connection, $pool);
+
+        $pool = $pools['pools-queue-queue']['pool'];
+        $connection = $pool->get();
+        $this->connections->add($connection, $pool);
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:31]: why are we adding the connection twice?

File: src/Appwrite/Platform/Tasks/ScheduleBase.php [hunk 34]
```diff
@@ -127,76 +128,74 @@ public function action(Group $pools, Database $dbForConsole, callable $getProjec
...
+                    }
+                }
+
+                $latestDocument = \end($results);
+            }
+
+            $lastSyncUpdate = $time;
+            $timerEnd = \microtime(true);
+
+            $connections->reclaim();
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/ScheduleBase.php:190]: where is connection defined ?

File: app/worker.php [hunk 35]
```diff
@@ -292,6 +121,13 @@
 }
 
 try {
+    $connection = new Connection\Redis(
+        System::getEnv('_APP_REDIS_HOST', 'redis'),
+        System::getEnv('_APP_REDIS_PORT', '6379'),
+        System::getEnv('_APP_REDIS_USER', ''),
+        System::getEnv('_APP_REDIS_PASS', '')
+    );
```
Comment (christyjacob4) [inline app/worker.php:129]: how will this work on cloud?

File: app/realtime.php [hunk 36]
```diff
@@ -348,13 +263,26 @@ function getCache(): Cache
...
-            $redis = $register->get('pools')->get('pubsub')->pop()->getResource(); /** @var Redis $redis */
+            $pools = $container->get('pools');
+            /** @var Connections $connections */
+            $connections = $container->get('connections');
+
+            $pool = $pools['pools-pubsub-pubsub']['pool'];
+            $connection = $pool->get();
+            $connections->add($connection, $pool);
+
+            $redis = $connection;
```
Comment (christyjacob4) [inline app/realtime.php:283]:
```suggestion
            $pools = $container->get('pools');
            $pool = $pools['pools-pubsub-pubsub']['pool'];
            
            /** @var Connections $connections */
            $connections = $container->get('connections');
            $connection = $pool->get();
            $connections->add($connection, $pool);

            $redis = $connection;
```
I would just rewrite this differently

File: app/realtime.php [hunk 37]
```diff
@@ -373,25 +301,24 @@ function getCache(): Cache
 
                     if ($realtime->hasSubscriber($projectId, 'user:' . $userId)) {
                         $connection = array_key_first(reset($realtime->subscriptions[$projectId]['user:' . $userId]));
-                        $consoleDatabase = getConsoleDB();
-                        $project = Authorization::skip(fn () => $consoleDatabase->getDocument('projects', $projectId));
-                        $database = getProjectDB($project);
+                        $consoleDatabase = $container->get('dbForConsole');
```
Comment (christyjacob4) [inline app/realtime.php:304]:
```suggestion
                        $dbForConsole = $container->get('dbForConsole');
```
lets call it this way

File: app/realtime.php [hunk 38]
```diff
@@ -373,25 +301,24 @@ function getCache(): Cache
...
+
+                        $project = $authorization->skip(fn () => $consoleDatabase->getDocument('projects', $projectId));
+                        $database = $container->get('getProjectDB')($project);
```
Comment (christyjacob4) [inline app/realtime.php:307]:
```suggestion
                        $dbForProject = $container->get('getProjectDB')($project);
```

File: app/realtime.php [hunk 39]
```diff
@@ -449,15 +386,16 @@ function getCache(): Cache
...
 
-        $dbForProject = getProjectDB($project);
-        $console = $app->getResource('console'); /** @var Document $console */
-        $user = $app->getResource('user'); /** @var Document $user */
-
+        $dbForProject = $container->get('getProjectDB')($project);
+        /** @var Document $console */
+        $console = $container->get('console');
+        /** @var Document $user */
+        $user = $container->refresh('user')->get('user');
```
Comment (christyjacob4) [inline app/realtime.php:398]: does `refresh()` make a new resource by making a new db query?

File: app/init.php [hunk 40]
```diff
@@ -1447,319 +781,614 @@ function (mixed $value) {
...
+    });
+
+$deviceForBuilds
+    ->setName('deviceForBuilds')
+    ->inject('project')
+    ->setCallback(function ($project) {
+        return getDevice(APP_STORAGE_BUILDS . '/app-' . $project->getId());
+    });
+
+$clients
```
Comment (christyjacob4) [inline app/init.php:1022]: these are all resources right? Wont it be good to move all resources to app/init/resources.php ?

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php (hunk 33 above)
Comment (byawitz) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:31]: The first one is for the `consoleDB`, and the second is for the `queue.`
I'll rename it for better clarity

File: src/Appwrite/Platform/Tasks/ScheduleBase.php (hunk 34 above)
Comment (byawitz) [reply src/Appwrite/Platform/Tasks/ScheduleBase.php:190]: Here https://github.com/appwrite/appwrite/blob/1581a8c654e17ab1f44673866f30c62a76157e8e/src/Appwrite/Platform/Tasks/ScheduleBase.php#L139

This is a local connections variable

File: app/worker.php (hunk 35 above)
Comment (byawitz) [reply app/worker.php:129]: With cloud specific overide

File: app/realtime.php (hunk 39 above)
Comment (byawitz) [reply app/realtime.php:398]: `refresh()` yes, it will force the re-execution of this 👇 function
https://github.com/appwrite/appwrite/blob/1581a8c654e17ab1f44673866f30c62a76157e8e/app/init.php#L553-L565
